File: webapp/src/lib/smallPart.ts
/**
 * The small-part solve: one net (or a pair) cut out over its planes, or a small drawn region.
 *
 * This is the browser's half of `worker/emi_worker/stages/small_part.py` and
 * `worker/emi_worker/openems/coupon.py`. The browser plans the coupon (its region and its
 * ports) and says what it will cost; the worker cuts the copper, meshes, and enforces the
 * budget against the real mesh. The two share their constants by hand, and a test pins them
 * (`smallPart.test.ts` against `server/emi/testdata/small_part_fixtures.json`).
 *
 * The cost shown here is an estimate from the region's size, fitted to coupons that were
 * actually meshed (docs/verification/small-part-solve.md). The worker publishes the real
 * mesh within seconds of starting and refuses a part its budget does not cover.
 */

import type { BoardDoc, BoardPad, GeometryIndex } from './boardTypes'
import { isReferenceNet, padLayer, type PortSpec } from './portPlacement'

export const SMALL_PART_MODE = 'small_part'

/** Mirrors of the worker's constants. Pinned by the shared fixture. */
export const MIN_MARGIN_MM = 3.0
export const MARGIN_HEIGHTS = 5.0
export const MAX_SIDE_MM = 60.0
export const PORT_HALF_WIDTH_MM = 0.2
export const MAX_CELLS = 3_000_000
export const MAX_CELL_STEPS = 1.5e11
export const MIN_RECORD_S = 10e-9
export const END_CRITERIA_DB = -50

/**
 * Cells per mm² of region by preset, and how much each copper layer past two adds, fitted to
 * meshed coupons; and the simulated time a terminated coupon took to ring down. Pinned by the
 * shared fixture and measured by `worker/research/sp_real_coupons.py`.
 */
export const CELL_FIT: Record<string, { per_mm2: number; per_layer: number }> = {
  coarse: { per_mm2: 500, per_layer: 0.8 },
  normal: { per_mm2: 1200, per_layer: 0.8 },
}
export const TYPICAL_RECORD_S = 5e-9

/** The worker's pessimistic throughput, MC/s. */
export const ASSUMED_MCELLS_S = 150

/**
 * Bands a user can pick. The first is the default: it keeps a terminated part's record to a
 * few nanoseconds while reaching past the fifth harmonic of most board clocks.
 */
export const BANDS = [
  { value: '100M-2G', label: '100 MHz to 2 GHz', lo: 100e6, hi: 2e9 },
  { value: '200M-1G', label: '200 MHz to 1 GHz', lo: 200e6, hi: 1e9 },
  { value: '300M-3G', label: '300 MHz to 3 GHz', lo: 300e6, hi: 3e9 },
] as const

export const PRESETS = [
  { value: 'coarse', label: 'Coarse', dx: 150, dy: 150, dz: 100 },
  { value: 'normal', label: 'Normal', dx: 75, dy: 75, dz: 50 },
] as const

export type Roi = [number, number, number, number]

/** Copper kept around the net, mm: max(3 mm, 5 heights of the outer dielectric). */
export function marginFor(doc: BoardDoc): number {
  const stack = doc.stackup.filter((s) => s.thickness_mm > 0 || s.role === 'copper')
  const copper = stack.map((s, i) => (s.role === 'copper' ? i : -1)).filter((i) => i >= 0)
  const outer: number[] = []
  if (copper.length) {
    for (const [k, step] of [[copper[0], 1], [copper[copper.length - 1], -1]] as const) {
      const j = k + step
      if (j >= 0 && j < stack.length && stack[j].role === 'dielectric') outer.push(stack[j].thickness_mm)
    }
  }
  const h = outer.length ? Math.max(...outer) : doc.board.thickness_mm
  return Math.max(MIN_MARGIN_MM, MARGIN_HEIGHTS * h)
}

/** The extent of some nets' copper, from the viewer's geometry plus pads and vias. */
export function netExtent(
  doc: BoardDoc, geometry: ArrayBuffer | null, index: GeometryIndex | null, nets: string[],
): Roi | null {
  const want = new Set(nets)
  let x0 = Infinity, y0 = Infinity, x1 = -Infinity, y1 = -Infinity
  const add = (x: number, y: number) => {
    if (x < x0) x0 = x
    if (x > x1) x1 = x
    if (y < y0) y0 = y
    if (y > y1) y1 = y
  }
  if (geometry && index) {
    const v = new Float32Array(geometry)
    for (const g of index.groups) {
      if (!want.has(g.net)) continue
      for (let k = g.offset; k < g.offset + g.count && 2 * k + 1 < v.length; k++) add(v[2 * k], v[2 * k + 1])
    }
  }
  for (const p of doc.pads) if (want.has(p.net)) add(p.x, p.y)
  for (const via of doc.vias) {
    if (!want.has(via.net)) continue
    const r = via.size_mm / 2
    add(via.x - r, via.y - r)
    add(via.x + r, via.y + r)
  }
  return Number.isFinite(x0) ? [x0, y0, x1, y1] : null
}

/** The other half of a differential pair, if the net's name says it has one. */
export function pairOf(doc: BoardDoc, net: string): string | null {
  const names = new Set(doc.nets.map((n) => n.name))
  for (const [a, b] of [['+', '-'], ['_P', '_N'], ['_p', '_n'], ['P', 'N']]) {
    for (const [from, to] of [[a, b], [b, a]]) {
      if (!net.endsWith(from)) continue
      const other = net.slice(0, -from.length) + to
      if (names.has(other)) return other
    }
  }
  return null
}

function pinCounts(doc: BoardDoc): Map<string, number> {
  const out = new Map<string, number>()
  for (const p of doc.pads) if (p.ref) out.set(p.ref, (out.get(p.ref) ?? 0) + 1)
  return out
}

/** The net's two ends: the pads furthest apart, the one on the part with most pins first. */
export function netEnds(doc: BoardDoc, net: string): BoardPad[] {
  const pads = doc.pads.filter((p) => p.net === net && p.type !== 'np_thru_hole')
  if (pads.length <= 1) return pads
  let best: [number, BoardPad, BoardPad] = [-1, pads[0], pads[1]]
  for (let i = 0; i < pads.length; i++) {
    for (let j = i + 1; j < pads.length; j++) {
      const d = Math.hypot(pads[i].x - pads[j].x, pads[i].y - pads[j].y)
      if (d > best[0]) best = [d, pads[i], pads[j]]
    }
  }
  const pins = pinCounts(doc)
  const key = (p: BoardPad) => [-(pins.get(p.ref) ?? 0), p.ref, p.number] as const
  return [best[1], best[2]].sort((a, b) => {
    const ka = key(a), kb = key(b)
    return ka[0] - kb[0] || ka[1].localeCompare(kb[1]) || ka[2].localeCompare(kb[2])
  })
}

export interface CouponPlan {
  nets: string[]
  roi: Roi | null
  ports: PortSpec[]
  margin_mm: number
  notes: string[]
  /** Why there is no coupon, in the user's words. */
  error: string | null
}

/** The coupon for these nets: its region and its ports, as the worker's `coupon.plan` makes it. */
export function planCoupon(
  doc: BoardDoc, geometry: ArrayBuffer | null, index: GeometryIndex | null, nets: string[],
): CouponPlan {
  const margin = marginFor(doc)
  const none = (error: string): CouponPlan => ({ nets, roi: null, ports: [], margin_mm: margin, notes: [], error })
  if (nets.length === 0) return none('Pick a net.')
  const ref = nets.find(isReferenceNet)
  if (ref) return none(`${ref} is a ground net. Pick the signal or supply that returns through it.`)
  const ext = netExtent(doc, geometry, index, nets)
  if (!ext) return none(`${nets.join(', ')} has no copper on this board.`)
  const roi: Roi = [ext[0] - margin, ext[1] - margin, ext[2] + margin, ext[3] + margin]
  const side = Math.max(roi[2] - roi[0], roi[3] - roi[1])
  if (side > MAX_SIDE_MM) {
    return none(
      `${nets.join(', ')} spans ${(side - 2 * margin).toFixed(0)} mm, a ${side.toFixed(0)} mm ` +
      `part with its planes. Small-part solves stop at ${MAX_SIDE_MM} mm. Pick a shorter net, ` +
      `or draw a region around the part you care about.`,
    )
  }
  const ports: PortSpec[] = []
  const notes: string[] = []
  for (const net of nets) {
    const ends = netEnds(doc, net)
    if (ends.length === 0) {
      notes.push(`${net} has no pads, so it has no port.`)
      continue
    }
    for (const pad of ends) {
      ports.push({
        name: `p${ports.length + 1}`, x_mm: pad.x, y_mm: pad.y, layer: padLayer(doc, pad),
        half_width_mm: PORT_HALF_WIDTH_MM, resistance: 50, excited: ports.length === 0,
        origin: 'net', net, padRef: pad.ref ? `${pad.ref}.${pad.number}` : undefined,
      })
    }
    if (ends.length === 1) notes.push(`${net} has one pad, so its far end is open.`)
  }
  if (ports.length === 0) return none(`${nets.join(', ')} has no pads to put a port on.`)
  return { nets, roi, ports, margin_mm: margin, notes, error: null }
}

export interface SmallPartEstimate {
  cells: number
  /** Smallest cell the mesh is expected to reach, mm, and the timestep it sets. */
  dt_seconds: number
  /** The timestep cap the worker will give it: three periods of the band's lowest frequency,
   *  or what the budget affords. */
  cap_steps: number
  /** Simulated time the cap allows, s. */
  record_s: number
  /** Wall time if the part rings down as the measured coupons did, and at the cap, s. */
  typical_seconds: number
  worst_seconds: number
  /** Why the worker would refuse it, or null. */
  refused: string | null
}

/**
 * What a part will cost. Cells come from a fit to meshed coupons (per preset, per mm² of
 * region, and per copper layer past two), and the timestep from the smallest cell the mesher
 * will make for a small part.
 */
export function estimateSmallPart(
  roi: Roi, preset: (typeof PRESETS)[number], band: (typeof BANDS)[number], doc: BoardDoc,
): SmallPartEstimate {
  const fit = CELL_FIT[preset.value]
  const area = Math.max(0, roi[2] - roi[0]) * Math.max(0, roi[3] - roi[1])
  const layers = doc.layers.length || 2
  const cells = Math.ceil(fit.per_mm2 * area * (1 + fit.per_layer * (layers - 2)))
  // The smallest cell: half of dx in plane (copper lines closer than that merge, and a routed
  // board always has some that close), or the thinnest slice of a dielectric at dz.
  const slices = doc.stackup
    .filter((s) => s.role === 'dielectric' && s.thickness_mm > 0)
    .map((s) => (s.thickness_mm * 1000) / Math.ceil((s.thickness_mm * 1000) / preset.dz))
  const dMin = Math.min(preset.dx / 2, preset.dz, ...slices) * 1e-6
  const dt = dMin / (299_792_458 * Math.sqrt(3))
  const needed = Math.ceil(3 / band.lo / dt)
  const affordable = Math.floor(MAX_CELL_STEPS / Math.max(cells, 1))
  const cap = Math.min(needed, affordable)
  const perStep = cells / (ASSUMED_MCELLS_S * 1e6)
  const typicalSteps = Math.min(cap, Math.ceil(TYPICAL_RECORD_S / dt))
  let refused: string | null = null
  if (cells > MAX_CELLS) {
    refused = `About ${fmtM(cells)} cells, over the ${fmtM(MAX_CELLS)} a small-part solve allows.`
  } else if (cap * dt < MIN_RECORD_S) {
    refused = `At about ${fmtM(cells)} cells the budget buys ${(cap * dt * 1e9).toFixed(1)} ns of ` +
      `simulated time; a part needs at least ${MIN_RECORD_S * 1e9} ns to ring down.`
  }
  return {
    cells, dt_seconds: dt, cap_steps: cap, record_s: cap * dt,
    typical_seconds: typicalSteps * perStep, worst_seconds: cap * perStep, refused,
  }
}

function fmtM(n: number): string {
  return n >= 1e6 ? `${(n / 1e6).toFixed(1)} M` : `${Math.round(n / 1e3)} k`
}

/** What the app sends for a small-part solve. The worker validates all of it again. */
export function smallPartParams(opts: {
  roi: Roi
  ports: PortSpec[]
  nets: string[]
  band: (typeof BANDS)[number]
  preset: (typeof PRESETS)[number]
  frequencies_hz: number[]
}) {
  const { roi, ports, nets, band, preset } = opts
  return {
    mode: SMALL_PART_MODE,
    roi: { min_x_mm: roi[0], min_y_mm: roi[1], max_x_mm: roi[2], max_y_mm: roi[3] },
    band_hz: [band.lo, band.hi],
    frequencies_hz: opts.frequencies_hz.filter((f) => f >= band.lo && f <= band.hi),
    mesh: { dx_um: preset.dx, dy_um: preset.dy, dz_um: preset.dz },
    ports: ports.map((p) => ({
      name: p.name, x_mm: p.x_mm, y_mm: p.y_mm, layer: p.layer,
      half_width_mm: p.half_width_mm, resistance: p.resistance, excited: p.excited,
      ...(p.net ? { net: p.net } : {}), ...(p.padRef ? { pad: p.padRef } : {}),
    })),
    ...(nets.length ? { coupon: { nets } } : {}),
  }
}

export function isSmallPartRun(params: unknown): boolean {
  return (params as { mode?: string } | undefined)?.mode === SMALL_PART_MODE
}

File: worker/emi_worker/stages/small_part.py
"""The bounded solve: one small part of a board, sized to finish in minutes.

A full-wave solve is expensive for reasons that have nothing to do with the part being asked
about. The timestep is set by the smallest cell anywhere, the record by the lowest frequency
(30 MHz needs 100 ns simulated), the far field adds a box a tenth of a wavelength out on every
side, and a board region full of other nets forces grid lines for all of them. A 15 mm region
of a real board with the far field took 2 h 38 min and left 57 of its 60 frequencies unusable.

This mode removes each of those on purpose:

* **A coupon, not a region** (``openems/coupon.py``): the net's copper over its planes, so the
  mesh is set by one net's edges, not by every trace near it. A region can still be drawn.
* **No far-field box.** The outputs are the near-field map and the port impedance and
  S-parameters, which need no air around the copper beyond the default.
* **A band that keeps the record short**: 100 MHz to 2 GHz by default. The run lasts until its
  energy has decayed ``END_CRITERIA`` below its peak, and with both ends of the net terminated
  in 50 ohm that is a few round trips of the coupon, not the 100 ns 30 MHz would ask for. The
  lower edge is what the excitation has to reach and what the timestep cap is sized from
  (three periods, 30 ns), so it sets the worst case; below 100 MHz a part a few centimetres
  across is quasi-static, and the answer is a lumped L or C a circuit tool gives cheaper. The
  upper edge keeps the coarsest cell under 3.4 mm in FR-4, which the default air and a coupon
  margin of a few millimetres hold without special meshing.
* **A lower end criterion**: -50 dB, where a solve otherwise stops at -40. A port read from a
  record cut at -40 dB is where the far field's negative resistances came from.
* **A budget checked against the real mesh**: cells times the timestep cap. A run that would
  exceed it is refused before it starts, with the numbers, rather than started and abandoned.

What is kept from every solve: a run that reaches its timestep cap publishes no number, one
that diverges is refused, and a frequency at which the port reads as a negative resistance is
dropped and listed in ``truncated_hz``.
"""

from __future__ import annotations

import json
import math

from ..openems import coupon as coupon_mod
from ..openems import hotspots as hotspots_mod
from ..openems import network as network_mod
from ..openems.model import SolveParams
from . import StageError

MODE = "small_part"

#: The default band, Hz. See the module docstring for why these two numbers.
BAND_HZ = (100e6, 2e9)

#: The widest band a small-part solve accepts, Hz. Below the floor the timestep cap (three
#: periods) outgrows any budget; above the ceiling a coupon's cells shrink with the wavelength
#: and the budget is spent on the mesh instead.
MIN_HZ = 50e6
MAX_HZ = 6e9

#: The energy decay a run stops at: -50 dB of its peak.
END_CRITERIA = 1e-5

#: Copper lines closer than this fraction of dx are one grid line (mesh.MERGE_FRACTION is a
#: quarter). The timestep is set by the smallest cell anywhere, and on a routed board a quarter of
#: dx is always reached: a pad corner beside a trace edge, an arc's vertices. A half moves no edge
#: by more than a quarter of dx, a detail the preset does not claim to resolve anyway, and on a
#: real 34 x 13 mm coupon it halved the cells and made the timestep 35 % longer (1.58 M cells
#: at 37.5 um to 0.80 M at 50.7 um, coarse).
MERGE_FRACTION = 0.5

#: The height every map is read at above its copper, mm, whatever the preset. The coarse
#: preset's first grid line above a trace is about this far up; a finer one reads closer, where
#: the field at a trace's edge is louder, and the presets disagreed by 1.7 dB on the level.
MAP_HEIGHT_MM = 0.1

#: A cell this coarse or coarser, um, is where a via's inductance was measured reading high: a
#: 0.3 mm via 1 mm from its return read +10.1 % on the coarse preset against the two-post
#: closed form, and +7.7 % on normal (docs/verification/small-part-solve.md, check 4). The
#: limit is 10 %, so a coarse part with vias says so rather than failing quietly.
COARSE_VIA_UM = 150.0

COARSE_VIA_NOTE = "coarse mesh: via inductance can read up to about 10% high"

#: The largest mesh a small-part solve will build, in cells (216 MB at 72 bytes a cell).
MAX_CELLS = 3_000_000

#: The most work a small-part solve may be asked for: cells times the timestep cap. At the
#: 150 MC/s the worker assumes, 1.5e11 is 17 minutes if the run went all the way to its cap;
#: a terminated coupon stops on its end criterion well before that.
MAX_CELL_STEPS = 1.5e11

#: The shortest record a small-part solve starts with, s. Below this the budget is too small
#: for the part to ring down, and the run is refused before it starts.
MIN_RECORD_S = 10e-9


def is_small_part(p: dict) -> bool:
    return str((p or {}).get("mode") or "") == MODE


def band(p: dict) -> tuple[float, float]:
    raw = p.get("band_hz")
    if not raw:
        return BAND_HZ
    try:
        lo, hi = float(raw[0]), float(raw[1])
    except (TypeError, ValueError, IndexError, KeyError) as exc:
        raise StageError("band_hz must be [low, high] in Hz") from exc
    if not (MIN_HZ <= lo < hi <= MAX_HZ) or hi < 2 * lo:
        raise StageError(
            f"a small-part band must lie within {MIN_HZ / 1e6:g} MHz to {MAX_HZ / 1e9:g} GHz "
            f"and span at least 2:1; got {lo / 1e6:g} to {hi / 1e6:g} MHz"
        )
    return lo, hi


def apply(p: dict, params: SolveParams) -> SolveParams:
    """Bound a solve's parameters to the small-part mode, or refuse what it cannot do."""
    for key, what in (("far_field", "a far field"), ("cable_ports", "cable ports"),
                      ("model_components", "component models")):
        if p.get(key):
            raise StageError(
                f"a small-part solve has no {what}. That needs a full-wave solve, which is a "
                f"separate experimental feature"
            )
    lo, hi = band(p)
    # Maps at the band's edges as well as wherever the user asked: the excitation is sized from
    # the lowest and highest dump frequency, and the port network is read across the whole band.
    asked = [f for f in params.frequencies_hz if lo <= f <= hi]
    freqs = sorted({lo, hi, *asked})
    params.frequencies_hz = freqs
    params.f_max = hi
    params.far_field = False
    params.cable_ports = {}
    params.model_components = False
    params.end_criteria = END_CRITERIA
    params.merge_fraction = MERGE_FRACTION
    # The band's coarsest cell holds everywhere, the absorbing layer included; the upper edge
    # of the band is chosen for it (module docstring), and the padding used to grow past it.
    params.pml_within_max_cell = True
    params.map_height_mm = MAP_HEIGHT_MM
    # The default band spans 20:1, just past where ``model.excitation_band`` warns that its ends
    # carry "a few dB more uncertainty", so every run said so. That is true of a level read
    # against the source, and nothing here is one: the port numbers are ratios of two transforms
    # of the same record, and the maps are per volt of the source. At 2 GHz the microstrip
    # check read Z0 within 0.5 % and S21 within 0.1 dB on every preset.
    params.band_edge_note = False
    # The cap is derived from the band (three periods of its lowest frequency); a hand-set cap
    # would make the budget below meaningless.
    params.max_timesteps = 0
    return params


def coupon_nets(p: dict) -> list[str]:
    c = p.get("coupon") or {}
    nets = c.get("nets") if isinstance(c, dict) else None
    if not nets:
        return []
    if not isinstance(nets, list) or not all(isinstance(n, str) and n for n in nets):
        raise StageError("coupon.nets must be a list of net names")
    if len(nets) > 4:
        raise StageError("a coupon takes at most four nets: a signal, its pair, and their return")
    return nets


def cut(board, transform, p: dict, params: SolveParams):
    """The board reduced to the coupon's copper, and what was said about it. Unchanged if the
    run named no nets (a drawn region)."""
    # A port the browser placed does not know what is under it: the browser has no pour
    # outlines. The worker does, so a port with no reference names the nearest layer with a
    # pour under it here, for a region as well as a coupon. Without one it falls back to the
    # nearest copper layer, as every solve always has, and the result says so.
    notes: list[str] = []
    for port in params.ports:
        if port.reference_layer:
            continue
        found = coupon_mod.reference_under(board, transform, port.x, port.y, port.layer)
        if found:
            port.reference_layer = found
        else:
            notes.append(
                f"there is no pour under port {port.name}, so it returns to the nearest copper "
                f"layer whether or not that layer has copper there. If the board has pours, "
                f"refill its zones in KiCad (B) before uploading: an unfilled zone is not copper"
            )
    nets = coupon_nets(p)
    if not nets:
        if _coarse(params) and _vias_in(board, transform, params.roi):
            notes.append(COARSE_VIA_NOTE)
        return board, notes
    missing = [n for n in nets if n not in set(board.nets) | {t.net for t in board.tracks}
               | {q.net for q in board.pads}]
    if missing:
        raise StageError(f"{', '.join(missing)} is not a net on this board")
    reduced, cut_notes = coupon_mod.extract(board, transform, nets, params.roi)
    cell = min(params.dx_um, params.dy_um) / 1000.0
    tight = coupon_mod.tight_gaps(reduced, nets, cell)
    if tight:
        many = tight != 1
        cut_notes.append(
            f"{tight} piece{'s' if many else ''} of other copper {'are' if many else 'is'} "
            f"closer to the net than one cell ({cell * 1000:.0f} um), so the mesh may join "
            f"{'them' if many else 'it'} to the net. A finer mesh preset resolves the gap"
        )
    if _coarse(params) and any(v.net in nets for v in reduced.vias):
        cut_notes.append(COARSE_VIA_NOTE)
    return reduced, notes + cut_notes


def _coarse(params: SolveParams) -> bool:
    return min(params.dx_um, params.dy_um) >= COARSE_VIA_UM


def _vias_in(board, transform, roi) -> bool:
    x0, y0, x1, y1 = roi
    return any(x0 <= x <= x1 and y0 <= y <= y1
               for x, y in (transform.pt(v.x, v.y) for v in board.vias))


def admit(cells: int, needed_steps: int, dt_seconds: float) -> int:
    """The timestep cap this run gets, or a refusal the user reads.

    The cap is the physics' own (three periods of the band's lowest frequency) or what the
    budget affords at this mesh, whichever is shorter. Three periods is a bound, not what a
    terminated coupon needs: the run ends on its energy criterion long before it, and a
    structure that has emptied has no spectrum left to truncate. So the budget shortens the
    cap rather than refusing outright, down to ``MIN_RECORD_S``; a run that then reaches the
    shortened cap is refused afterwards, like any run that reaches its cap, and says why.
    """
    if cells > MAX_CELLS:
        raise StageError(
            f"this part is too large for a small-part solve: {cells:,} cells, over the "
            f"{MAX_CELLS:,} it allows. Pick a shorter net, draw a smaller region or use the "
            f"coarse mesh"
        )
    affordable = int(MAX_CELL_STEPS // max(cells, 1))
    cap = min(needed_steps, affordable)
    if cap * dt_seconds < MIN_RECORD_S:
        raise StageError(
            f"this part is too large for a small-part solve: at {cells:,} cells its budget "
            f"buys {cap * dt_seconds * 1e9:.1f} ns of simulated time, and a part needs at least "
            f"{MIN_RECORD_S * 1e9:.0f} ns to ring down. Pick a shorter net, draw a smaller "
            f"region or use the coarse mesh"
        )
    return cap


def decay_fraction(energy_db: float) -> float:
    """How far a run's energy has fallen towards the end criterion, 0 to 1, for progress."""
    target = 10.0 * math.log10(END_CRITERIA)
    return min(1.0, max(0.0, energy_db / target))


def unusable_reason(result) -> str | None:
    """Why no number from this run may be shown, or None. The solver's own wording when it
    has one."""
    own = getattr(result, "unconverged_reason", None)
    if callable(own):
        return own()
    if result.converged:
        return None
    return (
        f"the run reached its limit of {result.max_timesteps:,} timesteps with its energy only "
        f"{abs(result.final_energy_db):.1f} dB down, before the fields settled, so no impedance "
        f"or S-parameter from it is used"
    )


def add_hotspots(artifacts, workdir: str, dump_names: dict[str, str],
                 dump_heights: dict[str, float], params: SolveParams, nearby) -> None:
    """List each map's separate spots within a few dB of its loudest (``openems.hotspots``),
    with the net and part nearest each, in the manifest. A failure leaves the list out."""
    import os

    from ..openems import post

    ref = artifacts.manifest.get("reference_magnitude") or 0.0
    if ref <= 0:
        return
    ports = [(q.x, q.y) for q in params.ports]
    floor = -float(artifacts.manifest.get("dynamic_range_db", post.DYNAMIC_RANGE_DB))
    out = []
    for layer, dump in dump_names.items():
        path = os.path.join(workdir, f"{dump}.h5")
        if not os.path.exists(path):
            continue
        try:
            grids = post.read_fd_dump(path, dump_heights.get(layer))
        except (OSError, ValueError):
            continue
        for g in grids:
            found = hotspots_mod.spots(g.x_mm, g.y_mm, g.to_db(reference=ref), ports, floor)
            for s in found:
                s["net"] = nearby.net(s["x_mm"], s["y_mm"], layer)
                s["part"] = nearby.part(s["x_mm"], s["y_mm"])
            if found:
                out.append({"layer": layer, "frequency_hz": g.frequency_hz, "spots": found})
    artifacts.manifest["hotspots"] = {"within_db": hotspots_mod.WITHIN_DB, "maps": out}
    artifacts.files["manifest.json"] = json.dumps(artifacts.manifest, indent=2).encode()


def add_network(artifacts, workdir: str, raw_ports: list[dict], params: SolveParams,
                unusable_reason: str | None) -> None:
    """Write ``network.json`` beside the maps. A failure leaves a note, not a failed run."""
    lo, hi = min(params.frequencies_hz), max(params.frequencies_hz)
    ports = []
    for q, raw in zip(params.ports, raw_ports):
        ports.append({"name": q.name, "resistance": q.resistance, "excited": q.excited,
                      "net": raw.get("net"), "pad": raw.get("pad")})
    try:
        doc = network_mod.network(workdir, ports, network_mod.log_grid(lo, hi),
                                  unusable_reason=unusable_reason)
    except (OSError, ValueError) as exc:
        artifacts.manifest["network_note"] = f"the port network could not be read ({exc})"
    else:
        artifacts.files["network.json"] = json.dumps(doc, separators=(",", ":")).encode()
        artifacts.manifest["network"] = {
            "format_version": network_mod.FORMAT_VERSION,
            "ports": [q["name"] for q in ports],
            "truncated_hz": doc["truncated_hz"],
        }
    artifacts.files["manifest.json"] = json.dumps(artifacts.manifest, indent=2).encode()
